check.sh

```diff
@@ -94,8 +56,6 @@ if git status --porcelain | grep -q "requirements.txt"; then
     echo "::endgroup::"
 fi
 
-codespell || EXIT_STATUS=$?
```

Comment:
(not the right line because github isn't letting me comment there)

Should you remove the `uv pip compile` lines above? And also update the error message + add a `pre-commit run -a` line in `check.sh` so people can run it locally? (I'm not sure people do run it locally but...)

Reply:
Realized we can't do `uv pip compile` in pre-commit in CI since it needs internet, so it needs to be here (or somewhere).

My goal is not to require/expect people to run check.sh locally as it's unintuitive and clunky, but given the limitations of pre-commit we do need to stash these somewhere and make it easy for users to run them locally if they want.
I'd personally vote for tox, and set up environments for the different tests.

Reply:
I usually wrap the pre-commit invocation in tox for local run anyway + run it in CI too, which covers the case of the service not having the internet during testing.

Reply:
We could also add a local system hook to pre-commit that runs `check.sh` if we were worried about people forgetting to run it

Reply:
> We could also add a local system hook to pre-commit that runs `check.sh` if we were worried about people forgetting to run it

There's several problems with doing this currently:
1. mypy is actually failing in my workdir, probably due to something dirty in my env. tox/nox would fix this
2. running `check.sh` takes a full minute on my current system, so I don't want to run it on every commit (`pre-commit run --all-files` takes ~8 seconds)
3. I catch most type errors from running `mypy` through pylsp in my editor anyway

So I don't ~ever run `check.sh` locally, and don't expect others to either. (Nor do I run `ci.sh`, I prefer manually running `pytest`)

This def isn't super obvious for new contributors though, but tox/nox should help with that. E.g. in tox you can create a label for a collection of environments that is considered "standard" to run locally.